### PR TITLE
Chapter 2

### PR DESCRIPTION
**Pull request by 7shantanu7**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -348,8 +359,10 @@ Implement a function that returns only the first half of a given list.
 >>> firstHalf "bca"
 "b"
 -}
--- PUT THE FUNCTION TYPE IN HERE
-firstHalf l = error "firstHalf: Not implemented!"
+firstHalf :: [a] -> [a]
+firstHalf l =
+    let halfLen = div (length l) 2
```
Comment (vrom911) [inline src/Chapter2.hs:364]: Nice usage of the `let` construction 👍🏼

File: src/Chapter2.hs [hunk 2]
```diff
@@ -842,7 +864,13 @@ list.
...
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate n l
+    | n < 0 = []
+    | null l = []
+    | otherwise =
+        let len = length l
+        in take len (drop (mod n len) (cycle l))
```
Comment (vrom911) [inline src/Chapter2.hs:873]: Cool optimisation 💯

**Review by vrom911 — APPROVED** [review]
Very well done! Awesome 👏🏼 💎

**Comment by chshersh** [comment]
We can't merge solutions to our repository, so I'm closing this PR.
Feel free to open PRs to unsolved chapters in your fork of this repository 🍴

**Pull request closed without merging**